### ISSUE TITLE
Fix "Maximum call stack size exceeded" errors from Google charts library

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#2247]
## Bug Description

Occasionally, any Google charts on the dashboard get replaced by "Maximum call stack size exceeded" errors. This happens in both the legacy content and the widgets API-driven content, so it's unrelated to that.

**Legacy UI:**
![Screenshot 2020-10-22 at 13 12 23](https://user-images.githubusercontent.com/3531426/96925120-85ddec00-1468-11eb-87ce-cab73a22bdd6.png)

**Widget API-driven UI:**
![Screenshot 2020-10-22 at 13 12 34](https://user-images.githubusercontent.com/3531426/96925133-8bd3cd00-1468-11eb-9b56-a27d6ff98bac.png)

It seems that this error happens primarily after returning to a previously opened tab with the Site Kit dashboard open. I'm not 100% about this assessment, but it has definitely _never_ happened to me on pageload (or more precisely, initial API response), always a bit later.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Google charts shouldn't occasionally be replaced by "Maximum call stack size exceeded" errors.

## Implementation Brief

This appears to be caused by us loading the chart library multiple times, best I could tell from similar errors encountered elsewhere. In the `develop` branch, the charts are loaded several times on the dashboard:

<img width="816" alt="Screenshot 2020-11-20 at 01 20 06" src="https://user-images.githubusercontent.com/90871/99743897-8ea8f880-2ace-11eb-8542-d25f94a0c7a9.png">

This seems intermittent (a race condition based on other instances of `GoogleChart`'s component/loading state).

* Ensure only one Google Charts loader script tag is loaded.
* Ensure `google.charts.load()` is only called once, and any remaining Charts instances wait for it to be ready.
* See #2408 for a proof-of-concept. (Rough, and could use improvements and possible datastore usage.)

Essentially, we should only be inserting one script tag into the DOM to load the Charts code and ensuring `google.charts.load()` is only called once.

Currently the proof-of-concept does this inside the component and by using the `wait-promise` module to check if `google.charts && google.charts.visualizations` is available. It could be worth storing the Google Charts loading state in the data store instead and providing a selector, as the existing code is not very easy-to-follow.

If no data store is to be used, we can tidy up the existing PoC (#2408) with better variable naming and reducing complexity where possible.

### Test Coverage

* None should be added; the E2E tests required would be quite slow and risk timeout. They would also be unreliable.

### Visual Regression Changes

* None.

## QA Brief

* Even with multiple refreshes and tabs open, Google Charts on the dashboard should not render a "Maximum call stack exceeded" error message.

## Changelog entry

* Fix component-specific initialization logic of Google charts library to prevent duplicate initialization.

**Comment by tofumatt** [comment]
This one turned out to be pretty tough to diagnose because it's intermittent and tough-to-reproduce, but I think I'm onto something… 🤞🏻 

After some web searching for similar issues and looking through the code, this might actually be related to us loading the charts library several times. I've seen some discussions on StackOverflow and Google Groups stating that calling the `.load` function several times can load the library a few times via a `<script>` tag. Additionally: our code would actually load the _loader_ multiple times because each component was trying to load it if `global.google?.charts` was false-y.

I think the solution here is to be better about managing our insertion of that script tag for the loader and about calling `google.charts.load()`. I think I have a solution, and it even means charts with pre-loaded data (like switching between date ranges) are more responsive (no loading animation):

![2020-11-19 21 23 17](https://user-images.githubusercontent.com/90871/99725786-7d032900-2aad-11eb-9a64-63ecc414490f.gif)

**Comment by felixarntz** [comment]
IB ✅ 

@tofumatt I think your approach sounds good. Let's not use the datastore for this one but solve it internally in the component, since IMO this is too much a technical component detail to have it covered in the datastore.

**Comment by aaemnnosttv** [comment]
@felixarntz @tofumatt – After seeing the PR for this, the problem seems to be deeper than a flaw in the Charts component. I'll create a new issue to address this as I already have a POC and we can make this issue dependent on it as it may no longer be relevant anymore.

**Comment by aaemnnosttv** [comment]
@tofumatt can you confirm if this is still an issue after #2444 ?

**Comment by tofumatt** [comment]
I tried to reproduce it but can't quite tell. I can't seem to get it to happen and to get the page loading more than one version of charts, but that's in part because the development page is where I encountered the error. The actual widget on that page is broken so it's hard to tell.

My thought is that the approach used to load a script here is still worthwhile, but after #2444 is merged and can be tested in `develop` then I can report further.

**Comment by aaemnnosttv** [comment]
Sounds good @tofumatt – let's make sure we can reproduce something about it after #2444 is merged – even if we can't reproduce the exact error, if we can still reproduce a condition that could result in charts loading multiple times then let's continue with this.

**Comment by aaemnnosttv** [comment]
@tofumatt This seems to be fixed by #2444. Adding a few logs in place I was able to reproduce the redundant loading of `loader.js` in develop but not in the branch for the multiple runtime fix.

<details>
	<summary>verbose details for nerds</summary>

```diff
diff --git a/assets/js/components/GoogleChart.js b/assets/js/components/GoogleChart.js
index 0e4610bad..2a27ed28f 100644
--- a/assets/js/components/GoogleChart.js
+++ b/assets/js/components/GoogleChart.js
@@ -46,6 +46,7 @@ async function loadCharts() {
 		script.type = 'text/javascript';
 
 		chartLoadPromise = new Promise( ( resolve ) => {
+			global.console.log( 'Loading Google charts loader.js' );
 			script.onload = resolve;
 			// Add the script to the DOM
 			global.document.head.appendChild( script );
@@ -81,6 +82,7 @@ export default function GoogleChart( props ) {
 	// Load the google charts library.
 	useEffect( () => {
 		loadCharts().then( () => {
+			global.console.log( 'Calling google.charts.load' );
 			global.google.charts.load( 'current', {
 				packages: [ 'corechart' ],
 				callback: () => {
```

**Dashboard on `develop`**
![image](https://user-images.githubusercontent.com/1621608/100845583-23740480-3486-11eb-8e9d-c1c25097e7e9.png)
![image](https://user-images.githubusercontent.com/1621608/100845594-266ef500-3486-11eb-92b7-6c905685a648.png)

**Dashboard on [`fix/2444-multiple-webpack-runtimes`](https://github.com/google/site-kit-wp/tree/fix/2444-multiple-webpack-runtimes)**
![image](https://user-images.githubusercontent.com/1621608/100845667-44d4f080-3486-11eb-857a-9f75a08797d6.png)
![image](https://user-images.githubusercontent.com/1621608/100845675-469eb400-3486-11eb-902a-29ac905e5c9e.png)

</details>


[According to the docs](https://developers.google.com/chart/interactive/docs/basic_load_libs#loading-details), it is okay to call `google.charts.load` more than once, but "safer not to do so" although it sounds like this is mostly only a concern if you're calling it with _different arguments_. Since there doesn't seem to be a good way to know if it's been called already or not (aside from maybe chart class presence?) I'm not sure there's much we can do about this.

I think the problem here is likely the multiple loaders as Google Charts should only be getting initialized on Site Kit screens by Site Kit, so there's little chance of it being called with different arguments. I would suggest we put this one on hold and see if any other issues come up. cc: @felixarntz

**Comment by tofumatt** [comment]
Unfortunately, it seems #2444 doesn't fix this issue.

Running the latest `develop` with a fresh build using `npm run watch`, I just got this error:

<img width="1392" alt="Screenshot 2020-12-04 at 22 03 59" src="https://user-images.githubusercontent.com/90871/101219934-aeb6eb00-367c-11eb-930b-9dc16df84744.png">

**Comment by tofumatt** [comment]
I'm just not able to reproduce the issue with my patch, so I think it's safer to go ahead with it. I've moved it back to code review 😄

**Comment by aaemnnosttv** [comment]
@tofumatt 

> Unfortunately, it seems #2444 doesn't fix this issue.
> Running the latest `develop` with a fresh build using `npm run watch`, I just got this error:

#2444 (PR #2445) hasn't been merged yet so that result would still be expected.

**Comment by cole10up** [comment]
### **QA ✅**

Ran a series of tests with multiple instances of the same page loaded on different browsers, same tab, adjusted the duration around on each and consistently refreshed over and over.

Left each idle for a bit and began refreshing multiple times.

![image](https://user-images.githubusercontent.com/57115907/101871500-84f83b00-3b38-11eb-8b9b-f3e178c4c5cc.png)

![Site Kit by Google Dashboard ‹ Hellogoodbye1 — WordPress (7)](https://user-images.githubusercontent.com/57115907/101873316-94c54e80-3b3b-11eb-872c-87b9475749f7.gif)

Unable to reproduce the issue. Sending to approval.
